### PR TITLE
adding a test for prototype of bound functions

### PR DESCRIPTION
**Pull request by claudepache**
Adding a test for the new initial value of the [[Prototype]] of bound functions, as discussed in Issue #553.

Also, add negative results for tested browsers (didn't test Edge).

### CODE REVIEW

**Comment by kangax** [comment]
@bterlson Brian, can you give result in Edge?

**Comment by webbedspace** [comment]
IMHO this should be categorised in "misc", where most of the other "ES5 compatibility breaks" are stored. ("Block-level function declaration (in strict mode)" isn't really a compatibility break since it wasn't available at all in ES5). Also unsure whether it should be its own test, or just stashed as a subtest of "miscellaneous".

File: data-es6.js [hunk 1]
```diff
@@ -2914,6 +2914,30 @@ exports.tests = [
   },
 },
 {
+  name: 'prototype of bound functions',
+  category: 'functions',
+  significance: 'small',
+  link: 'https://people.mozilla.org/~jorendorff/es6-draft.html#sec-boundfunctioncreate',
+  exec: function () {/*
+    function getProtoBound(proto) {
+      var f = Object.setPrototypeOf(function() { }, proto)
```
Comment (ljharb) [inline data-es6.js]: fwiw, this test will fail in any engine without `Object.setPrototypeOf` - is it worth creating a helper polyfill like the one in es6-shim, so that browsers with `__proto__` support but not `setPrototypeOf` can still pass this test (when testing shims, for example)?

File: data-es6.js [hunk 2]
```diff
@@ -2914,6 +2914,30 @@ exports.tests = [
...
+      var boundF = Function.prototype.bind.call(f, null)
+      return Object.getPrototypeOf(boundF)
+    }
+
+    return [ Function.prototype, function(){}, [], null ].every(function(proto) {
```
Comment (ljharb) [inline data-es6.js:2932]: It'd be nice, only in browsers that support arrow functions, to test an arrow function here too (and similarly, a generator function)

File: data-es6.js (hunk 1 above)
Comment (claudepache) [inline data-es6.js]: > fwiw, this test will fail in any engine without Object.setPrototypeOf - is it worth creating a helper polyfill like the one in es6-shim, so that browsers with **proto** support but not setPrototypeOf can still pass this test (when testing shims, for example)?

I don't think there is any engine that doesn't support setPrototypeOf but supports the new prototype initialisation of bound functions (which is a very recent change in the spec); therefore I didn’t care.

Comment (ljharb) [inline data-es6.js]: Right, but these tests need to run in older engines that have shimmed behavior. Since `Function#bind` is shimmable, the `es6-shim` will eventually correct this behavior, and would need to be able to pass in IE 9. Otherwise the test will only pass in browsers, never with shims/transpilers.

Comment (claudepache) [inline data-es6.js]: My guess is that `es6-shim` does already shim `Object.setPrototypeOf` now when possible, and won't be able to correct `Function#bind` otherwise. Unless you have a way to correct Function#bind without mutating the prototype?

File: data-es6.js (hunk 2 above)
Comment (claudepache) [inline data-es6.js:2932]: It seems to me that the utility is not worth the trouble of using `try/eval/catch` for the sake of older engines. (I do probably already too much by testing 4 possible prototypes.)

**Comment by claudepache** [comment]
> IMHO this should be categorised in "misc"

I think you are right. Are there opinions about whether to put it as a subtest of miscellaneous or not?

File: data-es6.js (hunk 1 above)
Comment (ljharb) [inline data-es6.js]: A ha, that's a fair point.

I suppose the only reason to care is if/when a browser fixes `bind` before implementing `setPrototypeOf`.

File: data-es6.js (hunk 2 above)
Comment (ljharb) [inline data-es6.js:2932]: `try`/`Function`/`catch` would be better, but you're right :-)

File: data-es6.js (hunk 1 above)
Comment (webbedspace) [inline data-es6.js]: My take on this is that one should make as few assumptions about what order platforms implement ES6 features as possible. So, I think this should, in the absence of `Object.setPrototypeOf`, fallback to the `__proto__` setter.

File: data-es6.js (hunk 2 above)
Comment (webbedspace) [inline data-es6.js:2932]: In my opinion, these should be tested, but included as separate subtests of this greater supertest (with the current test being moved to a "basic support" subtest).

Comment (claudepache) [inline data-es6.js:2932]: I fear that adding subtests for other types of functions (arrows, generators, subclasses, etc.) would add more distraction than useful information. Unless the tested implementation introduces a strange bug in its modified version of `.bind()`, the result of the subtest for `.bind` on generators would just be the intersection of the result of the base test for `.bind()` and the one for the existence of generators; and a negative result of that subtest could only be interpreted correctly when compared with the existence of generators.

Comment (kangax) [inline data-es6.js:2932]: I also think breaking those into subtests would be an overkill. Seems unlikely that one would work and not the other.

Comment (webbedspace) [inline data-es6.js:2932]: > Unless the tested implementation introduces a strange bug in its modified version of .bind()

Well, stranger bugs have happened. Consider the number of "instances of iterables" tests I've added to iterator consumer features - if they implemented the spec correctly, retrieving the `[Symbol.iterator]` method from the prototype chain should be no different than retrieving it directly. But, you can never be too sure. (In fact, I'm starting to wonder if I should split them all up into "generator instances" and "plain-object iterables" - [look at this in FF38](http://i.imgur.com/yGkJYPu.png)).

Comment (kangax) [inline data-es6.js:2932]: @webbedspace I trust your judgement then. If you think it's possible, let's do it that way.

Comment (claudepache) [inline data-es6.js:2932]: The issue with those subtests is that they may fail to give the wanted answer in a readable way.

As a developper, I am interested in: Does the implementation implements the correct semantics for bound functions, whatever their target function is?

Now, if I see the base test succeeding, but a subtest for some type of functions failing, I don't know the answer to my simple question, because I don't know if the subtest fails because of a bug in the implementation of `.bind`, or just because the implementation does not support that type of function.

(Note that there is currently a similar issue with the three last subtests of "__proto__ in object literals".)

I see two solutions, here:
- force a given subtest to pass if the implementation does not support the type of function that the subtest is testing; or,
- in addition to the "Yes"/"No" answers, add the possibility of a "N/A" answer.

Comment (ljharb) [inline data-es6.js:2932]: I would love all tests to provide an N/A answer when the thing they depend on (as opposed to the thing itself) is simply not supported yet in the engine. An alternative here would be for the subtest to pass - because arrow functions that aren't present don't fail, for example.

Comment (webbedspace) [inline data-es6.js:2932]: I don't see the necessity of the distinction. If you want to know "does the implementation implement the correct semantics for bound functions" then you need to assume generators exist, because the correct semantics for bound functions _require the existence of generators_. If generator functions do not exist, then any semantics involving them are obviously not implemented. (If you instead want to know "does the implementation implement the correct semantics for bound _normal_ functions" then the specific subtest provides complete information.)

I also don't see the necessity of the highly misleading "N/A" answer. If the bound generator functions test says "No", then you can find out whether basic generator support exists by checking the basic generator function test result. Either way, it simply tells the truth: no, the platform does not support feature X used with feature Y. Remember that in order to be ES6 compliant, _the entire feature set_, barring Annex B where noted, must be supported - not implementing every feature that intersects with a feature is leaving that feature incomplete.

File: data-es6.js [hunk 3]
```diff
@@ -2914,6 +2914,34 @@ exports.tests = [
...
+      return getProtoBound(proto) === proto
+    })
+  */},
+  res: {
+    ie10:        false,
+    firefox11:   false,
+    chrome:      false,
+    safari51:    false,
+    webkit:      false,
```
Comment (webbedspace) [inline data-es6.js:2941]: Just a note: every platform is assumed `false` unless `true` or something else is given. Therefore, listing any `false` values is redundant.

File: data-es6.js (hunk 2 above)
Comment (ljharb) [inline data-es6.js:2932]: I can see the argument for all of them. I'd say that if an engine doesn't implement, say, generators, then that doesn't mean that "all of the features that intersect with generators" either work, or don't work - it's simply not applicable.
